### PR TITLE
Implement DateIntervalFormatter

### PR DESCRIPTION
**Pull request by millenomi**
This is a bug-for-bug rendition of the equivalent Darwin class, plus a port of the relevant tests. Not all SPIs have been ported because they rely on ICU past 61.

### CODE REVIEW

**Comment by millenomi** [comment]
cc @spevans

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

File: CoreFoundation/Locale.subproj/CFDateIntervalFormatter.c [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+        switch (type) {
+            case kCFDateIntervalFormatterBoundaryStyleMinimizeAdjacentMonths:
+                uDateIntervalMinimizationStyle = UDTITVFMT_MINIMIZE_ADJACENT_MONTHS;
+                
+            default:
+                break;
+        }
+        
+        if (dif->_formatter) {
+            udtitvfmt_setAttribute(dif->_formatter, UDTITVFMT_MINIMIZE_TYPE, uDateIntervalMinimizationStyle, &status);
```
Comment (spevans) [inline CoreFoundation/Locale.subproj/CFDateIntervalFormatter.c]: Is this an Apple ICU added extra? It doesnt appear in the normal ICU headers

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by spevans** [comment]
I think the tests are failing because Apple's ICU uses a different separator to normal ICU, In my version I had to use the following in tests:

```
// Apple's ICU variant uses a THIN SPACE instead of a normal space around the separator.
#if canImport(Darwin)
    let separator = "\u{2009}\u{2013}\u{2009}"     // " – "  {THIN SPACE} {EN DASH} {THIN SPACE}
#else
    let separator = "\u{20}\u{2013}\u{20}"         // " – "  {SPACE} {EN DASH} {SPACE}
#endif
.
.
.
 XCTAssertEqual(dif.string(from: date1, to: date2), "18/12/2019, 20:00\(separator)19/12/2019, 22:15")
.
.
.
```
It's not obvious from the test log as they both look the same in monospaced. Maybe there is a way to set the separator character in the formatter to make them both the same?

**Comment by millenomi** [comment]
I test portions of the string now, to make sure the relevant data is or isn't there.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test and merge

**Comment by millenomi** [comment]
@swift-ci please test and merge

**Comment by ianpartridge** [comment]
Nice. Can we get this in Swift 5?

**Pull request merged**